### PR TITLE
Slim down Path::Combine by removing dependency on ValueTuple`8

### PR DESCRIPTION
**Pull request by marek-safar**
~- speeds up the implementation by 10%. It would be more if we could pass
the payload by reference~
- speeds up the implementation by 20% (for overload with 4 parameters)
- cuts SPC IL size by 12kB for most linked outputs

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.cs [hunk 1]
```diff
@@ -654,25 +654,47 @@ private static unsafe string JoinInternal(ReadOnlySpan<char> first, ReadOnlySpan
...
+                First = first;
+                FirstLength = firstLength;
+                Second = second;
+                SecondLength = secondLength;
+                Third = third;
+                ThirdLength = thirdLength;
+                Separators = separators;
+            }
+
+            public readonly IntPtr First;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.cs:681]: Use `char*` instead of `IntPtr` for all of these?

**Comment by jkotas** [comment]
> It would be more if we could pass the payload by reference

You should be able to pass the payload by reference by instantiating the factory over `IntPtr` and passing down the struct as pointer. It would actually reduce amount of unsafe code used in the implementation.

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.cs (hunk 1 above)
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/IO/Path.cs]: Done

**Comment by stephentoub** [comment]
> cuts SPC size by 12kB for most linked outputs

I assume that's for the crossgen'd image?

**Comment by marek-safar** [comment]
> Or is that simply how much IL there is associated with ValueTuple'8?

Yes, it's IL/metadata size for the type

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.cs [hunk 2]
```diff
@@ -654,83 +654,139 @@ private static unsafe string JoinInternal(ReadOnlySpan<char> first, ReadOnlySpan
...
                             destination[state.FirstLength] = PathInternal.DirectorySeparatorChar;
-                        new Span<char>((char*)state.Second, state.SecondLength).CopyTo(destination.Slice(state.FirstLength + (state.HasSeparator ? 0 : 1)));
+                        new Span<char>((char*)state.Second, state.SecondLength).CopyTo(destination.Slice(destination.Length - state.SecondLength));
                     });
             }
         }
 
+        private unsafe readonly struct Join3Payload
+        {
+            public Join3Payload (char* first, int firstLength, char* second, int secondLength, char* third, int thirdLength, byte separators)
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.cs:670]:
```suggestion
            public Join3Payload(char* first, int firstLength, char* second, int secondLength, char* third, int thirdLength, byte separators)
```

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.cs [hunk 3]
```diff
@@ -654,83 +654,139 @@ private static unsafe string JoinInternal(ReadOnlySpan<char> first, ReadOnlySpan
...
-            bool thirdHasSeparator = PathInternal.IsDirectorySeparator(second[second.Length - 1])
-                || PathInternal.IsDirectorySeparator(third[0]);
+            byte firstNeedsSeparator = PathInternal.IsDirectorySeparator(first[first.Length - 1])
+                || PathInternal.IsDirectorySeparator(second[0]) ? 0 : 1;
+            byte secondNeedsSeparator = PathInternal.IsDirectorySeparator(second[second.Length - 1])
+                || PathInternal.IsDirectorySeparator(third[0]) ? 0 : 1;
 
             fixed (char* f = &MemoryMarshal.GetReference(first), s = &MemoryMarshal.GetReference(second), t = &MemoryMarshal.GetReference(third))
             {
+                var payload = new Join3Payload (
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.cs:701]:
```suggestion
                var payload = new Join3Payload(
```

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.cs [hunk 4]
```diff
@@ -654,83 +654,139 @@ private static unsafe string JoinInternal(ReadOnlySpan<char> first, ReadOnlySpan
...
                             destination[destination.Length - state.ThirdLength - 1] = PathInternal.DirectorySeparatorChar;
-                        new Span<char>((char*)state.Third, state.ThirdLength).CopyTo(destination.Slice(destination.Length - state.ThirdLength));
+                        new Span<char>(state.Third, state.ThirdLength).CopyTo(destination.Slice(destination.Length - state.ThirdLength));
                     });
             }
         }
 
+        private unsafe readonly struct Join4Payload
+        {
+            public Join4Payload (char* first, int firstLength, char* second, int secondLength, char* third, int thirdLength, char* fourth, int fourthLength, byte separators)
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.cs:727]:
```suggestion
            public Join4Payload(char* first, int firstLength, char* second, int secondLength, char* third, int thirdLength, char* fourth, int fourthLength, byte separators)
```

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.cs [hunk 5]
```diff
@@ -654,83 +654,139 @@ private static unsafe string JoinInternal(ReadOnlySpan<char> first, ReadOnlySpan
...
+            byte firstNeedsSeparator = PathInternal.IsDirectorySeparator(first[first.Length - 1])
+                || PathInternal.IsDirectorySeparator(second[0]) ? 0 : 1;
+            byte secondNeedsSeparator = PathInternal.IsDirectorySeparator(second[second.Length - 1])
+                || PathInternal.IsDirectorySeparator(third[0]) ? 0 : 1;
+            byte thirdNeedsSeparator = PathInternal.IsDirectorySeparator(third[third.Length - 1])
+                || PathInternal.IsDirectorySeparator(fourth[0]) ? 0 : 1;
 
             fixed (char* f = &MemoryMarshal.GetReference(first), s = &MemoryMarshal.GetReference(second), t = &MemoryMarshal.GetReference(third), u = &MemoryMarshal.GetReference(fourth))
             {
+                var payload = new Join4Payload (
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.cs:764]:
```suggestion
                var payload = new Join4Payload(
```

**Comment by jkotas** [comment]
> cuts SPC size by 12kB for most linked outputs

Just curious - where is Join with 3 or 4 arguments used in trivial apps?

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.cs [hunk 6]
```diff
@@ -654,83 +654,139 @@ private static unsafe string JoinInternal(ReadOnlySpan<char> first, ReadOnlySpan
                 return string.Create(
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IO/Path.cs:654]: This one should be smaller and faster as:
```
return hasSeparator ? string.Concat(first, second) : string.Concat(first, PathInternal.DirectorySeparatorCharAsString, second);
```
(The existing implementation may need to stay for `MS_IO_REDIST`.)

**Review by jkotas — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/IO/Path.cs (hunk 6 above)
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/IO/Path.cs]: Good suggestion!

**Comment by marek-safar** [comment]
> where is Path.Join with 3 or 4 arguments used in trivial apps?

Overload with 3 arguments (which before pulled ValueTuple`8)

https://github.com/dotnet/runtime/blob/8fc4859e1037170f1b207a700d6dbda269c90550/src/libraries/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs#L734

**Pull request merged**